Adds Collections API, makes menus use it

hypeJunction opened the pull request.

hypeJunction committed:
feat(util): adds API for managing collections of items

Adds API for creating array-like collections of unique items

hypeJunction committed:
feat(menus): wrap menu items as collections

Menus service now wraps menus, menu section and menu items as collections,
to make it easer to alter them without having to loop every time an
item is needed.

jeabakker commented on engine/classes/Elgg/Collections/Collection.php:
```diff
@@ -0,0 +1,270 @@
+			return $this;
+		}
+
+		$this->items[$item->getId()] = $item;
```
If this collection is constructed with `\ElggObject` as the item class this will fail.

Eighter add a check here that $item implements `CollectionItemInterface` or make the constructor through an exception if `$item_class` doesn't implement `CollectionItemInterface`

hypeJunction replied:
I was throwing an exception but then PHPStorm keeps highlight stuff with Unhandled exception. But yes, you have a point.

hypeJunction committed:
chore(tests): define testing strategy for menu hook testing

Defines a general approach to testing menu hooks and adds tests for
users related menus

hypeJunction commented:
As a next step, I would want to extend collections API to support entities, especially plugins. Public DI container will get a record for `plugins`, which by default will query the database for active plugins, but the boot service will be able to substitute the value, so instead of messing with the plugins table service, we will just maintain a separate collection of active plugins.

jeabakker commented on engine/classes/Elgg/Menu/MenuItems.php:
```diff
@@ -0,0 +1,12 @@
+/**
+ * A collection of menu items
+ */
+class MenuItems extends Collection {
```
wouldn't it make sense to overrule the `__construct()` and default the `$item_class` to `ElggMenuItem::class`?? It could save some weird constructor calls below

hypeJunction replied:
Sure.

hypeJunction replied:
Done

jeabakker commented on engine/classes/Elgg/Menu/Service.php:
```diff
@@ -141,11 +146,12 @@ function combineMenus(array $names = [], array $params = [], $new_name = '') {
-		$prepared_items = [];
-		
+	public function prepareMenuItems($items = []) {
+		$prepared_items = new MenuItems([], ElggMenuItem::class);
```
like here

jeabakker commented on engine/classes/Elgg/Menu/UnpreparedMenu.php:
```diff
@@ -23,15 +25,23 @@ class UnpreparedMenu {
 		$this->params = $params;
+
+		if (is_array($items)) {
+			$items = new MenuItems($items, ElggMenuItem::class);
```
and here

jdalsem commented:
Do you also think of hooks/events lists as a target for this API?

hypeJunction commented:
Anything that uses arrays can potentially be transitioned. A variation of this would be turning $vars into an object and removing all the boilerplate around extracting, unsettling, pushing, merging

mrclay commented on engine/classes/Elgg/Collections/Collection.php:
```diff
@@ -0,0 +1,284 @@
+			$map[$id] = call_user_func($callback, $item, $id);
+		}
+
+		return $map;
```
filter returns a collection but not map?

hypeJunction replied:
Mapped items do not necessarily represent a required item instance class

mrclay commented:
I don't wanna throw water on this, but [doctrine/collections](https://www.doctrine-project.org/api/collections/latest/Doctrine/Common/Collections/ArrayCollection.html) is already in our lock file.

mrclay commented:
For $vars, [ParameterBag](https://api.symfony.com/2.3/Symfony/Component/HttpFoundation/ParameterBag.html) makes a lot of sense, but I'd be hesitant to add even more overhead to views.

hypeJunction commented:
Ah, I didn't know about doctrine collections. We can extend that class.
I see the point about $vars, we would need to do some benchmarking. It's not something that can be done easily anyway, at first we need to stop type hinting arrays and allow array objects all over the API. Something to consider for the future

hypeJunction commented:
I can't make this work with doctrine's collections.

hypeJunction commented:
Closing this. If anyone cares enough to rewrite with Doctrine's collections, be my guest.
